src/test/util/setup_common.cpp

```diff
@@ -120,18 +120,61 @@ BasicTestingSetup::BasicTestingSetup(const ChainType chainType, const std::vecto
         arguments = Cat(arguments, G_TEST_COMMAND_LINE_ARGUMENTS());
     }
     util::ThreadRename("test");
-    fs::create_directories(m_path_root);
-    m_args.ForceSetArg("-datadir", fs::PathToString(m_path_root));
-    gArgs.ForceSetArg("-datadir", fs::PathToString(m_path_root));
     gArgs.ClearPathCache();
     {
         SetupServerArgs(*m_node.args);
+        m_node.args->AddArg("-testdatadir", "Custom data directory (default: <random_string>)", ArgsManager::ALLOW_ANY, OptionsCategory::DEBUG_TEST);
         std::string error;
         if (!m_node.args->ParseParameters(arguments.size(), arguments.data(), error)) {
             m_node.args->ClearArgs();
             throw std::runtime_error{error};
         }
     }
+
+    if (!m_node.args->IsArgSet("-testdatadir")) {
+        // By default, the data directory has a random name
+        const std::string testdatadir{g_insecure_rand_ctx_temp_path.rand256().ToString().c_str()};
+        m_path_root = fs::temp_directory_path() / "test_common_" PACKAGE_NAME / testdatadir;
+        fs::create_directories(m_path_root);
+    } else {
+        // Custom data directory
+        m_has_custom_datadir = true;
+        const std::string testdatadir{m_node.args->GetArg("-testdatadir").value()};
+        m_path_root = fs::path(testdatadir.c_str()) / "test_temp";
+
+        // Try to obtain the lock; if unsuccessful don't disturb the existing test.
+        auto lock_datadir = [&]() -> void {
+            try {
+                fs::create_directories(m_path_root);
+            } catch (const std::exception&) {
+            }
+            if (!fs::exists(m_path_root)) {
+                std::cerr << "Cannot create test data directory " + fs::PathToString(m_path_root) + '\n';
+                exit(1);
```

Comment:
This should use the `EXIT_FAILURE` constant, not exit(1).

Also, I would move the print + exit to a standalone function to dedup code. Like `Shutdown(const std::string& err_msg);`.